app/src/main/java/com/example/perfectweatherallyear/MainActivity.kt

```diff
@@ -19,21 +30,61 @@ class MainActivity : AppCompatActivity() {
 
         firebaseAnalytics = Firebase.analytics
 
-//        binding.crashButton.setOnClickListener {
-//            firebaseAnalytics.logEvent(FirebaseAnalytics.Event.SELECT_ITEM) {
-//                param(FirebaseAnalytics.Param.ITEM_ID, binding.crashButton.id.toString())
-//                param(FirebaseAnalytics.Param.ITEM_NAME, "CrashButton")
-//            }
-//            Firebase.crashlytics.log("this is my runtime exception")
-//            Firebase.crashlytics.setCustomKeys {
-//                key("my_log", "this is my runtime exception")
-//            }
-            //throw RuntimeException("Test Crash") // Force a crash
-//        }
+        getDynamicLinkFromFirebase()
+        openLinkFromNotification()
+    }
+
+    private fun openLinkFromNotification() {
+        val link = intent.getStringExtra("link")
+        if (link != null) {
+            val uri = Uri.parse(link)
+            getParamAndOpenDeepLink(uri)
+        }
+    }
+
+    private fun getDynamicLinkFromFirebase() {
+        Firebase.dynamicLinks.getDynamicLink(intent)
+            .addOnSuccessListener(this) { pendingDynamicLinkData: PendingDynamicLinkData? ->
+                var deepLink: Uri? = null
+
+                if (pendingDynamicLinkData != null) {
+                    deepLink = pendingDynamicLinkData.link
+
+                    if (deepLink != null) {
+                        getParamAndOpenDeepLink(deepLink)
+                    }
+                }
+            }
+            .addOnFailureListener(this) {
+                    e -> Log.i("DynamicLink", "getDynamicLink:onFailure", e) }
+    }
+
+    private fun getParamAndOpenDeepLink(uri: Uri) {
+        val city = uri.getQueryParameter(CITY_NAME_PARAM)
+        val id = uri.getQueryParameter(CITY_ID_PARAM)?.toInt()
+
+        if (id != null && city != null) {
+            val location = Location(id, city)
+            addFragment(WeatherForecastFragment(), location)
+        }
     }
 
     override fun onStop() {
         super.onStop()
         ForecastNotifierWorker.createWorker(this)
         }
+
+    private fun addFragment(fragment: Fragment, location: Location) {
```

Comment:
Why you don't use navigation component for navigation? You can [handle](https://developer.android.com/guide/navigation/navigation-deep-link) deep link  navigation by navigation component